index.js

```diff
@@ -82,14 +89,72 @@ app.post('/check', async (req, res) => {
 
     let signature = omeOrder.signed_data.toString()
 
-    let isValidSig = validateSignature(contractOrder.order, process.env.TRADER_CONTRACT, network, signature)
-    let isWhitelisted = validateWhitelist(web3, contractOrder.order.maker)
-    
-    if (isValidSig && isWhitelisted) {
-        return res.status(200).send()
-    } else {
-        return res.status(400).send()
+    const isValidSig = validateSignature(contractOrder.order, process.env.TRADER_CONTRACT, network, signature)
+    if(!isValidSig) {
+        return res.status(400).send({
+            message: 'Invalid signature, ensure the correct Trading Contract is being used',
+            order: contractOrder.order
+        });
     }
+
+    const isWhitelisted = validateWhitelist(web3, contractOrder.order.maker)
+    if(!isWhitelisted) {
+        return res.status(400).send({
+            message: 'Address not whitelisted for trading',
+            order: contractOrder.order
+        });
+    }
+
+    // validate timestamps
+    const isValidCreatedTime = validateCreatedTime(contractOrder.order.created)
+    if(!isValidCreatedTime) {
+        return res.status(400).send({
+            message: 'Order created time is invalid',
+            order: contractOrder.order
+        });
+    }
+
+    const isValidExpiryTime = validateExpiryTime(contractOrder.order.expires)
+    if(!isValidExpiryTime) {
+        return res.status(400).send({
+            message: 'Order expiry time is invalid',
+            order: contractOrder.order
+        });
+    }
+
+    // check if user has enough margin in the market
+    const perpetualSwapContract = new web3.eth.Contract(perpetualSwapABI, contractOrder.order.market)
+
+    const feeRate = await perpetualSwapContract.methods.feeRate().call()
+    const maxLeverage = await perpetualSwapContract.methods.maxLeverage().call()
+
+    const currentPosition = await perpetualSwapContract.methods.getBalance(contractOrder.order.maker).call()
```

Comment:
Might be premature optimisation, but we could cache these?

Early on we will only have 1 or 2 markets so loading in and reading these at each and every trade submission seems cumberson.

Probably scope creep on this ticket, but worth doing at some point in the future I'd say

Reply:
its a good point, can `feeRate` and `maxLeverage` change once the contract is deployed?

Reply:
@raymogg it looks like those two variables can be updated by the contract owner (not sure what that actually looks like in practice), In order to cache these effectively we would need a reliable way to invalidate the cache if these variables were to change on the contract.

I'll make a ticket to address the cache invalidation issue and proceed with this in its current state if you are cool with that?

Reply:
ticketed up here: https://mycelium-eth.atlassian.net/browse/LMDEVS-712